internal/check/engine.go

```diff
@@ -33,91 +40,475 @@ func NewEngine(d EngineDependencies) *Engine {
-			return false, err
+		// an empty page token denotes the first page (as tokens are opaque)
+		var prevPage string
+
+		// Special case: check if we can find the subject id directly
+		if rels, _, err := e.d.RelationTupleManager().GetRelationTuples(ctx, requested.ToQuery()); err == nil && len(rels) > 0 {
+			resultCh <- checkgroup.Result{
+				Membership: checkgroup.IsMember,
+				Tree: &expand.Tree{
+					Type:  expand.Leaf,
+					Tuple: requested,
+				},
+			}
+			return
 		}
 
-		allowed, err := e.subjectIsAllowed(ctx, requested, nextRels, restDepth)
+		g := checkgroup.New(ctx)
+
+		for {
+			nextRels, nextPage, err := e.d.RelationTupleManager().GetRelationTuples(ctx, expandQuery, x.WithToken(prevPage))
+			// herodot.ErrNotFound occurs when the namespace is unknown
+			if errors.Is(err, herodot.ErrNotFound) {
+				g.Add(checkgroup.NotMemberFunc)
+				break
+			} else if err != nil {
+				g.Add(checkgroup.ErrorFunc(err))
+				break
+			}
 
-		// loop through pages until either allowed, end of pages, or an error occurred
-		if allowed || nextPage == "" || err != nil {
-			return allowed, err
+			g.Add(e.checkDirect(ctx, requested, nextRels, restDepth-1))
+
+			// loop through pages until either allowed, end of pages, or an error occurred
+			if nextPage == "" || g.Done() {
+				break
+			}
+			prevPage = nextPage
 		}
 
-		prevPage = nextPage
+		resultCh <- g.Result()
 	}
 }
 
-func (e *Engine) SubjectIsAllowed(ctx context.Context, r *relationtuple.InternalRelationTuple, restDepth int) (bool, error) {
-	// global max-depth takes precedence when it is the lesser or if the request max-depth is less than or equal to 0
+func (e *Engine) SubjectIsAllowed(ctx context.Context, r *RelationTuple, restDepth int) (bool, error) {
+	result := e.Check(ctx, r, restDepth)
+	return result.Membership == checkgroup.IsMember, result.Err
+}
+
+func (e *Engine) Check(ctx context.Context, r *RelationTuple, restDepth int) checkgroup.Result {
+	// global max-depth takes precedence when it is the lesser or if the request
+	// max-depth is less than or equal to 0
 	if globalMaxDepth := e.d.Config(ctx).MaxReadDepth(); restDepth <= 0 || globalMaxDepth < restDepth {
 		restDepth = globalMaxDepth
 	}
 
-	return e.checkOneIndirectionFurther(ctx, r, &relationtuple.RelationQuery{Object: r.Object, Relation: r.Relation, Namespace: r.Namespace}, restDepth)
+	resultCh := make(chan checkgroup.Result)
+	go e.checkIsAllowed(ctx, r, restDepth)(ctx, resultCh)
+	select {
+	case result := <-resultCh:
+		return result
+	case <-ctx.Done():
+		return checkgroup.Result{Err: context.Canceled}
+	}
+}
+
+func (e *Engine) checkIsAllowed(ctx context.Context, r *RelationTuple, restDepth int) checkgroup.Func {
+	if restDepth < 0 {
+		e.d.Logger().
+			Debug("reached max-depth, therefore this query will not be further expanded")
+		return func(_ context.Context, resultCh chan<- checkgroup.Result) {
+			resultCh <- checkgroup.Result{Membership: checkgroup.MembershipUnknown}
```

Comment:
Each check function writes to the resultCh exactly once, and the caller must ensure that the result is read. If the context is cancelled, the check function must write `Result{Err: ctx.Err()}` to the channel.

Not receiving from the channel would be a resource leak. In https://github.com/ory/keto/blob/a400364880fe6f34c89c8b90424ad529d74c15dc/internal/check/checkgroup/concurrent_checkgroup.go#L52 we explicitly receive all running go routines.

Alternatively, we could change the semantics to say that the check function must not send on the `resultCh` if the `ctx` was cancelled. WDYT?

Reply:
I since discovered go.uber.org/goleak, which checks that there are no more goroutines running at the end of a test. I'll add this, so that we can be sure that nothing leaks or hangs.

Reply:
Done, assumptions are now thoroughly tested :).